### ISSUE TITLE
Proposition for a more flexible corpora structure

### ISSUE THREAD

**Issue by bguil** [UniversalDependencies/docs#1146]
In the UniDive task 1.5 (dedicated to improving the consistency of guidelines and annotations for spoken data across UD treebanks), with @kajad, @ellepannitto  and @sylvainkahane, we would like to suggest a more flexible organisation of CoNLL-U files within each UD treebank.

We propose that each treebank can choose how to split its data into individual CoNLL-U files. Spoken treebanks are usually composed of a set of audio recordings and a structure in which each CoNLL-U file corresponds to one recording is a very natural organisation. This is also because, somehow differently from at least some written resources, the order in which sentences appear is crucial to the structure of the resource, and splitting in dev, test, train based on the number of tokens could break the usability of the data. Many existing treebanks have a such an organisation, which is stored in the `not-to-release` folder (English-EWT, etc.) or elsewhere in the corpus specific management infrastructure (English-GUM, most SUD treebanks, Italian-Valico, etc.).

It is important to note that any existing treebank could keep its existing structure for the `dev`, `test` and `train` files without making any changes.

If we want to keep the `dev`, `test` and `train` split available for all treebanks, we can easily come up with a way to encode it (for instance, the [one used by EWT](https://github.com/UniversalDependencies/UD_English-EWT/tree/dev/not-to-release/file-lists), or just adding `# train`, `# dev` and `# test` as metadata on the sentences).

We could also decide on a conventional naming for an implicit sentence-level metadata which refers to the filename (something like `sample_id`, `document_id` or `file_id`). This could be used in a query language (such as Grew-Match) to restrict queries to sub-corpora or to create clusters according to the different sub-corpora.

**Comment by dan-zeman** [comment]
As with everything in the core of the infrastructure, I tend to be very conservative with respect to introducing changes. I also do not find one of the motivations very strong - you definitely should not split a document between train/dev/test (I think the split guidelines actually discourage you from doing so). Inside one file, document boundaries should be marked if available, so extraction of subcorpora should be possible even now (but I think we should improve the level of standardization _how_ various attributes of documents are encoded so that the selection of documents for a particular purpose is easier).

That said, I agree that having more than 3 CoNLL-U files can be useful, and we were actually forced to do it with a few treebanks where the training file is larger than what GitHub allows. Right now these are exceptions hard-coded in a script. Perhaps making this part of the system rather than an exception would help, as it would increase the flexibility significantly. But I would keep the strict naming conventions we have now (which includes the labels `train`, `dev` or `test` in the file name).

**Comment by sylvainkahane** [comment]
It would be interesting to know who are our users today. I believe that the proportion of (computational) linguists working with the UD collection constantly increase and should be in the majority. For sure the proportion of users that need the dev train test is in the minority and they are people that can easily reorganize a treebank as they want. It is why we recommend to have the most visible organization of a UD treebank following the standards in corpus construction. It is very inconvenient to maintain the resource with one organization and to have to change it at each release to another.

**Comment by amir-zeldes** [comment]
I agree with Dan that changing the format in repos is tricky, especially because, as @sylvainkahane points out the UD user base is always growing and already quite big. Whenever some change like this is implemented, hundreds of scripts written by grad students everywhere can stop working. If they are still there, they need to fix them, and if they graduate, it can be very hard to figure out what is broken and how to fix it, so I prefer to keep things stable if there is a good way to do so.

If the issue is aligning files to recordings, I think you could probably use conllu comments to indicate whenever the file changes? And if you still want to include conllu files per recording, you can use the `not-to-release` folder you mentioned. So it seems to me that the current structure can still work.

**Comment by ellepannitto** [comment]
Why should the scripts break (and keeping things as they are for the sake
of maintaining not-so-good code is anyway not the best reason in my
opinion)? Already existing resources can keep the tree file organization if
they find it convenient.

 We're discussing about allowing also a different organization for new
resources for which the train-dev-test is just a mostly useless wrapping
that has to be created only for the purpose of the release...

Il Lun 29 Set 2025, 23:36 Amir Zeldes ***@***.***> ha scritto:

**Comment by nschneid** [comment]
One thing to keep in mind is that some systems train a parsing model on each UD treebank after each release. (I'm thinking of UDPipe and Stanza, though there may be others.) We should make sure any change in filenames won't cause invalid training or invalid reported scores of such systems.

**Comment by Stormur** [comment]
One thing that would be good to introduce is a complete conllu file by default, something like `ud-xxx_all.conllu`. I always create such a file anyhow when investigating something. Also the default creation of a big aggregated file per language could be useful. This would need some more standards on sentence identifiers, though.

**Comment by amir-zeldes** [comment]
> Why should the scripts break

That depends on what the scripts do :) I think training a parser on a list of UD languages as mentioned by @nschneid is a plausible scenario, and running a search engine on freshly pulled data is another. So is getting a lexicon out of corpus data for a range of languages, and in many cases people would leave out 'test' because they want to be able to get scores on whatever they're doing. 

But I don't pretend to know all or even a majority of the things people do with UD datasets. My point is that when a project is widely used in the community, making breaking changes should be avoided unless there is a strong need, and even then figuring out ways to be backwards compatible is important. If we ever release a UD v3 that would be a good time to make necessary breaking changes. Consider that just the docs repo has 262 forks and 289 stars, so I think it's not an exaggeration to say that hundreds of people regularly use UD repos - changing the file structure could affect lots of running systems and regular update scripts. Linking to audio files that are smaller than a split partition could be achieved under the current standard using comment lines.

> the order in which sentences appear is crucial to the structure of the resource, and splitting in dev, test, train based on the number of tokens could break the usability of the data. 

I forgot to comment on this point - I agree with @dan-zeman completely that data from the same document should ideally not be in two different partitions, and would probably inflate scores, since there will be many more coincidental similarities between sentences in the same document and for example parsers will perform better on those parts that are in test than they would on genuinely unseen data (sometimes these overlaps might be unavoidable, but at least as an aspiration). I will also add that most written treebanks do preserve contiguous text and sentence ordering, and that some spoken ones also don't for various reasons. I don't think using multiple files per partition should change our ability to ensure that data is sequential within files - that is also already possible using the current format.

**Comment by martinpopel** [comment]
The (computational) linguists working with UD I know appreciate the simplicity of just three files. Most of them even understand the importance of train/dev/test splits and know they should base (parameters of) their linguistic hypotheses on train only (and validate on dev or dev+test). And I hope all of them know how to concatenate the three files if they need it.

> complete conllu file by default, something like `ud-xxx_all.conllu`

I vote against this proposal. It would increase the risk that someone trains on test sets. Doing `cat *.conllu` is a matter of one second, but when I see that I am concatenating train+dev+test, the risk of unintentional misuse is much lower.

I understand that the original proposal was to keep the train/dev/test split (but only for existing treebanks!) and have the "more flexible organisation" only in addition to the existing one. However, the above-mentioned risk of misuse would be still high if the there is train+dev+test mixed together in the official UD release.

Maybe we could provide a script, which produces one file per document. I usually use `cat *.conllu | udapy read.Conllu split_docs=1 write.Conllu docname_as_file=1` for such purpose, but we could have something more robust (working even for files where only `newdoc` is present, but without the document name; placing the new files into train/dev/test directories).

In addition, we could add a requirement (to the validator, but first as a warning) that each treebank must use `newdoc id = xy` and the document names (`xy`) are valid file names on Unix&Windows. I know I am twisting this issue from "more flexible" to "more strict", but I think it would be more useful for all the users (be them computational linguists or not) and just slightly more work for the treebank maintainers.

**Comment by ellepannitto** [comment]
I think that the point is that no one that deals (in this specific case, but it also applies to more broad scenarios) with spoken data has their data in the train-dev-test splits while annotating, and adding document metadata in the conllu doesn't solve that because no software that allows to annotate is able to keep track of that kind of metadata.
So the reality becomes that all of these resources are developed somewhere else, in other repositories, with slightly different formats and organizations, and then need to be translated into the train-dev-test split to be released.

This, in my opinion, is not the best for reproducibility because the versioned data is not the actual data but just a projection of it, and it leaves behind a lot of the great features of working on github such as using the issues and discussing in-between releases and being able to see what is happening on dev.

It is very easy, if you're training a model, to get yourself the split (even just a command like the one @martinpopel is suggesting), since you're anyway already scripting to do that (and you have to do some sort of selection anyway, for instance to separate text genres or documents if you want to train on a subset of the data), but it might not be as easy for the linguists that are actually annotating and curating the data to adapt to train-dev-test setting, in some subfields more than in others.

**Comment by Stormur** [comment]
> I vote against this proposal. It would increase the risk that someone trains on test sets. Doing `cat *.conllu` is a matter of one second, but when I see that I am concatenating train+dev+test, the risk of unintentional misuse is much lower.

I do not understand this. The risk is already there and if anybody is doing evaluation cycles they will be aware of the issue. 

> Most of them even understand the importance of train/dev/test splits and know they should base (parameters of) their linguistic hypotheses on train only (and validate on dev or dev+test).

Also this: why? If I want to investigate linguistic patterns, I need all available data. Maybe we are not thinking about the same things?

> I usually use `cat *.conllu | udapy read.Conllu split_docs=1 write.Conllu docname_as_file=1` for such purpose

So this is not trivial. Providing the script is a good move. But at that point, it might be more useful to provide a script to perform the train/test/dev split on a complete file, even if it probably is more complex! This would imply indeed a more standardised way to delimit specific subcorpora.

**Comment by dan-zeman** [comment]
The reason why the UD guidelines always required a pre-defined (and ideally stable across releases) train-dev-test split is reproducibility and comparability. If people report parsing scores, we want to be able to say that the scores have been measured on the same test set (and also that the models were trained on the same training set). At present, the UD release number is enough to know that models are comparable. If everyone wrote their own script to extract their own test data, comparability would be gone.

I agree that looking for linguistic patterns can be usually safe on concatenation of the parts (if one can guarantee that the results of such studies will not be used in parser development), and for smaller treebanks it would be unfortunate to not look at half of the data just because it is used to test parsers. But it still has to be labeled as test data, so that people who train anything know that they must not look at it.

And while the above may sound like something that does not bring any benefit to linguists, in reality it does: Linguists form a significant group of users of pretrained parsers such as UDPipe and Stanza. So the necessity for @foxik to change UDPipe training scripts if the file structure changes is only part of the problem. The potentially bigger problem is that if your treebank does not have training data, no models pretrained on your treebank will be made available. The technically capable users may be able to train a parser themselves, but the purely linguistic users of parsers are the ones that will suffer.

That said, you can actually even now opt-out from splitting the data and keep everything labeled as test. A few treebanks have chosen to do so, although they are large enough to be able to be split. By default the validation infrastructure will bug you that you have forgotten to split the dataset to train(-dev)-test, so I would have to register an exception on demand (and I would try to talk you out of it in case yours is the only treebank of a given language), but it si doable.

> no software that allows to annotate is able to keep track of that kind of metadata

In that case, changing the UD infrastructure is not the only possible remedy; one could instead fix the software :-)

**Comment by ellepannitto** [comment]
> In that case, changing the UD infrastructure is not the only possible remedy; one could instead fix the software :-)

Then I would propose to enforce a very strict and transparent mapping between sent_id and document_id, since those are anyway only relevant to the UD split. Because otherwise fixing all the softwares that do not use the only python library that is able to handle document level metadata in conllu is also a huge amount of work...

**Comment by amir-zeldes** [comment]
> transparent mapping between sent_id and document_id

I think that's a good idea to recommend, and at least for datasets developed at Georgetown the sentence IDs indeed include the document identifier. They are also sequential following the original order after each new document ID, and I think that should be done if it's possible in a dataset (though sometimes it is not).

> fixing all the softwares

Actually my point above was that changing the structure could be harmful to software. UD has had this file structure for a long time, so any software developer who wanted to do something with UD repos has had a long time to realize that and write appropriate code. Changing the structure is more likely to be harmful to software than not changing it.

> complete conllu file by default, something like ud-xxx_all.conllu

Like @martinpopel , I think if we allow this it's almost inevitable that someone will glob("*.conllu") on a UD repo at some point and create results with duplicate sentences.

**Comment by ellepannitto** [comment]
Yes I was also thinking about something like "if there is a document id in the file, then all sentences id should be prefixed by the document id". My concern is about softwares like ELAN or all of the stuff people that work with spoken/first hand data work with, and need to be able to import and export data without losing the document structure just because the document level metadata gets lost in this file organization

**Comment by dan-zeman** [comment]
> > In that case, changing the UD infrastructure is not the only possible remedy; one could instead fix the software :-)
> 
> Then I would propose to enforce a very strict and transparent mapping between sent_id and document_id, since those are anyway only relevant to the UD split. Because otherwise fixing all the softwares that do not use the only python library that is able to handle document level metadata in conllu is also a huge amount of work...

The CoNLL-U file format does not define document-level metadata, at least not formally. Even the optional document id is formally just another sentence-level comment. And so are various other pieces of information that logically pertain to the whole document and that people want to store in the treebank: typically they are stored at the first sentence of the document. Of course it could be improved but that seems to be a different issue than whether treebanks should have the train, dev and test file, so maybe we should make it a separate thread. At any rate, sentence-level comments are firm part of CoNLL-U / UD, and UD-capable annotation tools should be able at least to preserve them, regardless of which libraries they use.

Repeating document id as a prefix of sentence id is a practice I would also recommend, although UD guidelines do not require it (not least because they do not require document ids). If it helps you organize stuff in your favorite annotation tool, even better.

**Comment by ellepannitto** [comment]
I think everybody is using "good practices", like also @amir-zeldes was saying their corpora include document ID in the sentence IDs and I know that @bguil and @sylvainkahane have a document-like organization for SUD, and I'm sure that also others had nice and useful ideas... 

The point of the whole issue, I think, is finding a way to make people that work on a document level organization more "natively" able to work in the UD framework, instead of asking them to translate their work for every release.
One way could be changing file organization, another could be changing the way metadata is enforced. I'm not necessarily leaning for one of the two, but I think that introducing a _validated_ / _enforced_ / _true_ standard about this among all UD treebanks rather than just a "good practice" would make resources more easy to develop within the framework and more comparable and would pave the way for the development of a lot more stuff that is not only "training on train without looking into it and testing on test", which is just one of the way models can be developed on UD resources, maybe not the most useful one for people working on interactional and spoken data.

**Comment by amir-zeldes** [comment]
Yes, I would also welcome a standard on document and (sub)corpus-level metadata. There is some recent discussion in #1135 .

**Comment by dan-zeman** [comment]
I'm not trying to dismiss the discussion, but I'm a bit skeptical about enabling everybody work "more natively" in UD. UD is not a general-purpose annotation framework; its files can hold quite a lot of additional annotation but they are not optimized towards this extra stuff. People have different needs and priorities, sometimes going in opposite ways; if their priorities do not align well with those of the UD framework, conversion is inevitable. In the first years of UD, all treebanks were converted from a non-UD framework. And some of them are still maintained in the source format and converted to UD if there are changes upstream.

**Comment by harisont** [comment]
I just read the whole discussion and it seems that there are good arguments for both altering and preserving the current three-file structure. If train-dev-test splits are maintained, however, I have a side note on this (emphasis mine):

> you can actually even now opt-out from splitting the data and keep everything labeled as test. A few treebanks have chosen to do so, _although they are large enough to be able to be split_

I really like that UD encourages incremental releases, and I think having a lower bound when it comes to splitting is problematic in this sense. Let's say my treebank initially contains 10K words. As of now, I am forced to release them as a test set. I see two potential problems:

1. by the time the treebank grows to splitting size, the distribution of the data might have changed, and the optimal train/dev/test splitting might involve reshuffling sentences so that the content of the test split change, creating a problem in terms of reproducibility
2. some of my users might want to decide to use (some of) my 10K sentences for training anyway and report results on a fraction of the test set. With fine-tunable models, the amount of data that makes a treebank relevant for some kind of training is changing, so why not allow train-dev-test splitting regardless the size of the treebank? I think treebank developers can be trusted to know what people may want to do with their data and to ask for advice if in doubt.

**Comment by dan-zeman** [comment]
> 1. by the time the treebank grows to splitting size, the distribution of the data might have changed, and the optimal train/dev/test splitting might involve reshuffling sentences so that the content of the test split change, creating a problem in terms of reproducibility

While we try to minimize the chance of ill-designed experiments by discouraging people from changing the train-dev-test split without good reasons, I would say that this is one of the possible good reasons; and I also think that any published results should always clearly indicate which UD release they were trained and evaluated on (and never mix them). As a reviewer I would object if a paper just says "we train on Universal Dependencies".

> 2. some of my users might want to decide to use (some of) my 10K sentences for training anyway and report results on a fraction of the test set. With fine-tunable models, the amount of data that makes a treebank relevant for some kind of training is changing, so why not allow train-dev-test splitting regardless the size of the treebank? I think treebank developers can be trusted to know what people may want to do with their data and to ask for advice if in doubt.

The [recommended sizes](https://universaldependencies.org/contributing/repository_files.html#data-split) were introduced before the CoNLL shared task in UD parsing in 2017 (it included only treebanks that had at least 10K words in test data). The recommendation stays there to encourage similar design when possible (i.e., this is the advice that people may ask for). The recommendation also says that _"with small treebanks people should do ten-fold cross-validation"_ – so training+testing on such treebanks is possible, we just do not have enough data to make the ideal experimental setting possible.

Finally, it is actually possible to split a small treebank and we have some (e.g., UD_Marathi-UFAL has 2997 training words, 438 dev, and 412 test words; it was first published in fall 2017, the author knew that he could not participate in the 2018 shared task anyway (having only 3847 words in total), so he split what he had). It is true that in some configuration the validation will report cases of not-following the split recommendations, without emphasizing that there are other options; then the people get in touch and obtain a registered exception. But for treebanks that have less than 10K words, validation should not complain about the split at all.

**Comment by harisont** [comment]
Thank you for lots of useful information! I suppose I gave the recommendations too much weight, and I was unfamiliar with the existing exceptions. 

Glad to see both of the issues I raised seem to be non-problems after all :)

**Comment by dan-zeman** [comment]
> I gave the recommendations too much weight, and I was unfamiliar with the existing exceptions

Not announcing them too loudly is intentional :-)
